numpy/typing/tests/test_typing_extensions.py

```diff
@@ -0,0 +1,56 @@
+"""Tests for the optional typing-extensions dependency."""
+
+import sys
+import types
+import inspect
+import importlib
+
+import typing_extensions
+import numpy.typing as npt
+
+
+def _is_sub_module(obj: object) -> bool:
+    """Check if `obj` is a `numpy.typing` submodule."""
+    return inspect.ismodule(obj) and obj.__name__.startswith("numpy.typing")
+
+
+def _is_dunder(name: str) -> bool:
+    """Check whether `name` is a dunder."""
+    return name.startswith("__") and name.endswith("__")
+
+
+def _clear_attr(module: types.ModuleType) -> None:
+    """Clear all (non-dunder) module-level attributes."""
+    del_names = [name for name in vars(module) if not _is_dunder(name)]
+    for name in del_names:
+        delattr(module, name)
+
+
+MODULES = {"numpy.typing": npt}
+MODULES.update({
+    f"numpy.typing.{k}": v for k, v in vars(npt).items() if _is_sub_module(v)
+})
+
+
+def test_no_typing_extensions() -> None:
+    """Import `numpy.typing` in the absence of typing-extensions.
+
+    Notes
+    -----
+    Ideally, we'd just run the normal typing tests in an environment where
+    typing-extensions is not installed, but unfortunatelly this is currently
+    impossible as it is an indirect hard dependency of pytest.
+
+    """
+    assert "typing_extensions" in sys.modules
+
+    try:
+        sys.modules["typing_extensions"] = None
+        for name, module in MODULES.items():
+            _clear_attr(module)
+            assert importlib.reload(module), name
+    finally:
+        sys.modules["typing_extensions"] = typing_extensions
+        for module in MODULES.values():
+            _clear_attr(module)
+            importlib.reload(module)
```

Comment:
Admitetly this is more of a hack then I'd ideally like it to be.
If anyone has suggestions for an alternative approach, then that would be very much welcome.

Reply:
Yeah, this is pretty terrifying :).  We have a `test_full_reimport` which does the same thing in a subprocess.  Can you do that instead?  If I understand it right, using `sys.modules["typing_extensions"] = None` is basically the "magic" ingredient here.  Moving it into a subprocess will remove the complexity of reimporting numpy proper.

Reply:
>  If I understand it right, using `sys.modules["typing_extensions"] = None` is basically the "magic" ingredient here

Yes, by assigning an invalid value to the module an `ImportError` will be triggered whenever
executing `import typing_extensions` statements (which is the desired outcome here). 

I'm gona do a bit of experimenting with this `subprocess`-based approach, see if I can get it to work.

Reply:
Alright, changed as of https://github.com/numpy/numpy/pull/19525/commits/ee9855823c41c0f5d4d2aa81e1965e898a375fbe.

I also pushed a small fix for `test_full_reimport` (https://github.com/numpy/numpy/pull/19525/commits/1ff314dc000cb0a7a1bb2823d3bc3ac4171f9d16). Namelly, exceptions raised within a subprocess 
stay within the subprocess, so they're effectively suppressed unless you actually validate the return code.